### PR TITLE
codeintel: Add max age to non-relevant failed index records

### LINKED ISSUE #40417: Expiration of failed upload and index records
We currently expire completed upload records to remove large associated data from the codeintel-db. We have no data retention on failed upload records, or data retention on index metadata records (at all). The lsif_upload tables can become quite big, which is bad for some classes of SQL queries that currently use table scans.

We should add a background job that removes lsif_upload records that:
- have failed
- are older that some configurable threshold (default, say, ~1 month?)
- are not the last record of its type (for a repo/root)

We should also add a background job that removes lsif_index records in a similar fashion.

### PR DESCRIPTION
**Pull request by efritz**
Fixes #40417. This PR adds a new query to the autoindexing cleanup routine that occurs periodically. We now remove old (~3 months) and failed indexing records that would not be the most recent failure for a line of indexing (same repo, root, indexer).

## Test plan

New unit tests.

### CODE REVIEW

File: internal/codeintel/autoindexing/internal/store/store_expiration.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+	return (s.db.Exec(ctx, sqlf.Sprintf(expireFailedRecordsQuery, now, int(failedIndexMaxAge/time.Second), batchSize)))
+}
+
+const expireFailedRecordsQuery = `
+WITH
+ranked_indexes AS (
+	SELECT
+		u.*,
+		RANK() OVER (PARTITION BY repository_id, root, indexer ORDER BY finished_at DESC) AS rank
```
Comment (Strum355) [inline internal/codeintel/autoindexing/internal/store/store_expiration.go:30]: Does the order of columns matter here for the index?

File: internal/codeintel/autoindexing/internal/store/store_expiration.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	FROM lsif_indexes u
+	WHERE
+		u.state = 'failed' AND
```
Comment (Strum355) [inline internal/codeintel/autoindexing/internal/store/store_expiration.go:33]: Im confused on "failed" vs "errored", i think workers.md accidentally misuses one at some stage

File: internal/codeintel/autoindexing/internal/store/store_expiration.go (hunk 1 above)
Comment (efritz) [reply internal/codeintel/autoindexing/internal/store/store_expiration.go:30]: No, I think the query planner would be able to shuffle these as they're only used for uniqueness. However, query plans show that it's more advantageous to grab all failed indexes and quicksort to partition.

File: internal/codeintel/autoindexing/internal/store/store_expiration.go (hunk 2 above)
Comment (efritz) [reply internal/codeintel/autoindexing/internal/store/store_expiration.go:33]: cc @eseliger

File: internal/codeintel/autoindexing/internal/store/store_expiration.go (hunk 1 above)
Comment (Strum355) [reply internal/codeintel/autoindexing/internal/store/store_expiration.go:30]: COOL. At most maybe newline split this for readability 🙂

File: internal/codeintel/autoindexing/internal/store/store_expiration.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+		%s - u.finished_at >= %s * interval '1 second'
+),
+locked_indexes AS (
+	SELECT i.id
+	FROM lsif_indexes i
+	JOIN ranked_indexes ri ON ri.id = i.id
+	WHERE ri.rank != 1 OR EXISTS (
```
[the remark this answers is not in the document]
Comment (efritz) [reply internal/codeintel/autoindexing/internal/store/store_expiration.go:49]: Done.

**Review by Numbers88s — APPROVED** [review]

**Pull request merged**